Fix saga

gcperk opened the pull request:
This branch was largely an update of saga scripts (landscape and model covars) to streamline and remove the unwanted sysdata.rda and external data files. 
I added two internal check function to connect to saga (find_saga_path() and check_saga().  These are currently internal, however they might need to be fully documented as the reference within the create_covariates link is showing up as a Warning?
```<html>
<body>
<!--StartFragment-->
❯ checking Rd cross-references ... WARNING   Missing link or links in documentation object 'create_covariates.Rd':     'find_saga_path'      Missing link or links in documentation object 'create_samplr_covariates.Rd':     'find_saga_path'      See section 'Cross-references' in the 'Writing R Extensions' manual.  0 errors ✔ \| 1 warning ✖ \| 0 notes ✔
--
 
> | >
>

<!--EndFragment-->
</body>
</html>

 ```
This pull request will close #13 . 

I added basic test functionality, but interested on your assessment to add a snapshot to test running the entire suite of outputs. 
I also suspect the internal point to a local .saga exe is causing the tests to fail, so commenting out for now. Is there a nice way to deal with this? 
ie avoid: sp <- "C:/Programs/saga-9.2.0_x64/saga-9.2.0_x64/saga_cmd.exe" within the testing.

Happy to get your views on overall functionality and any improvements .

gcperk commented:
@ateucher - I think this is ready for review, before I get too far down the track. It is throwing an apple/mac error still but everything else is passing. Happy to chat first if that is easier. 
Thanks!

ateucher commented:
Thanks @gcperk! I've started figuring out how to install on a Mac - notes here: https://github.com/ninoxconsulting/PEMprepr/issues/10#issuecomment-2455955103. I think I'm almost there and can then test this :)

Also in that comment is a link to install it on Linux, so we could do that in GitHub actions and test conditionally there...

gcperk commented:
thanks @ateucher. I'm not totally clear how much of the setup is required to get the tests to successfully run? Do you need anything from me at the moment?

ateucher commented on tests/testthat/test-create_landscape_covariates.R:
```diff
@@ -0,0 +1,12 @@
+#   outdir <- withr::local_tempdir()
+#   sp <- "C:/Programs/saga-9.2.0_x64/saga-9.2.0_x64/saga_cmd.exe"
+#   aoi_snapped <- make_test_aoi(outdir)
+#   aoi_rast <- create_template_raster(aoi_snapped, res = 50, out_dir = outdir)
```
I think these arguments need to be updated: aoi -> dtm

gcperk replied:
thanks - I updated these and added a placeholder for correct working function output

ateucher commented on tests/testthat/test-create_landscape_covariates.R:
```diff
@@ -0,0 +1,12 @@
+#   expect_error(create_landscape_covariates(aoi =  aoi_rast , saga_path = NULL))
+#
+# })
+
```
I know they're not actually running yet, but some tests to test the functionality will be necessary I think (i.e., not just errors)

ateucher replied:
You could even just put them in as placeholders for now until we get the SAGA detection working

ateucher commented on R/create_landscape_covariates.R:
```diff
@@ -0,0 +1,345 @@
+    ridge <- (tp > 1)
+    ridge[stats::na.omit(ridge)] <- 6
+
+    # plot(valley, main="Valley", sub="TPI <= -1", cex.main=0.9, cex.sub=0.7, legend=FALSE)
```
At some point it would be good to move this commented code out of here and into a script in a `scratch` folder or something like that, it makes it a bit hard to navigate...

ateucher commented on R/create_landscape_covariates.R:
```diff
@@ -0,0 +1,345 @@
+
+  #--- SAGA ---#
+
+  if (!is.null(saga_path)) {
+    check_saga(saga_path)
+    cli::cli_alert_success(
+      "Your SAGA connection has been succesfuly set up. Using the {.var {saga_path}}"
+    )
+  } else {
+    check_saga()
+    cli::cli_abort("{.var saga_path} must point to the saga_path location on your computer.
+                   Please check your program files or equivalent to locate the saga_path.exe file")
+  }
```
I think this can be greatly condensed, since this logic is already in `check_saga()`

```suggestion
check_saga(saga_path)
```

ateucher replied:
Ditto for the same code in `create_covariates()`

ateucher commented on R/create_covariates.R:
```diff
@@ -1,217 +1,189 @@
+  # layers = "all"
+  # tile = FALSE
 
+  moddir <- fs::path_package("PEMprepr", "extdata/saga_module_depends.csv")
+  # moddir <- read.csv("saga_module_depends.csv")
```
```suggestion
  moddir <- read.csv(fs::path_package("PEMprepr", "extdata/saga_module_depends.csv"))
```

Line 118 currently calling `filter()` on a file path, need `read.csv()`

gcperk replied:
Shifted this to line 118

ateucher commented on R/create_covariates.R:
```diff
@@ -1,217 +1,189 @@
-  if(Sys.info()['sysname']=="Windows"){
-    saga_cmd <- file.path(saga_path, "saga_cmd.exe")
-    fns      <- "\\" ### file name separator
+  if (!is.null(saga_path)) {
```
Ditto below - can replace L65-74 with `check_saga()`

ateucher reviewed (commented):
Looking good! Not a full review yet, but `create_covariates()` and `create_landscape_covariates()` are very long, with some bits that seem like they repeat similar patterns, similar to `create_base_vectors()`.

A few comments inline for now; I'm going to work on some infrastructure to detect and use SAGA in different situations...

gcperk commented:
Great thanks @ateucher, we completely agree they are long!! I will wait til you are finished with the entire review before I make any changes. A major challenge I have is with the testing as it relies on the saga_path so could see that being an issue. Any ideas there? Thanks for the review as always.

ateucher commented on inst/extdata/saga_module_depends.csv:
```diff
@@ -0,0 +1,47 @@
+19,tpi,,,
+20,ridgevalley,sinksfilled,,
+21,mrn,sinksfilled,,
+22,flowaccumulation,sinksfilled,,TRUE
+23,flowaccumulation,sinksfilled,,TRUE
```
@gcperk are these rows meant to be duplicated?

gcperk replied:
Yes - I deleted one row

ateucher commented on R/create_covariates.R:
```diff
@@ -1,217 +1,189 @@
+  deps <- setdiff(unique(c(moddirs$dep1, moddirs$dep2)), moddirs$module)
+  deps <- deps[!grepl("^\\s*$", deps)]
+  if (length(deps > 0)) {
+    mods <- c(moddir$module, deps)
```
Should this be `moddirs`, i.e., the filtered data frame? I think so...

```suggestion
    mods <- c(moddirs$module, deps)
```

ateucher commented on R/create_covariates.R:
```diff
@@ -1,217 +1,189 @@
+    cli::cli_abort("{.var layers} must be a character string of the names of variables")
   }
 
   if (isTRUE(layers == "all")) {
     layers <- layer_options
   }
```
```suggestion
  if (isTRUE(layers == "all")) {
    layers <- layer_options
  } else {
    layers <- c("dem_preproc", layers)
  }
```

ateucher replied:
This would be adding dem_preproc, as it seems to be an unstated prerequisite for many

ateucher commented on R/create_covariates.R:
```diff
@@ -1,217 +1,189 @@
 
 
+  # order the modules to be run
+  layers_to_call <- mods[order(match(mods, moddir$module))]
```
I got duplicates in some cases

```suggestion
  layers_to_call <- unique(mods[order(match(mods, moddir$module))])
```

ateucher commented on R/create_covariates.R:
```diff
@@ -1,217 +1,189 @@
 #' #--- create all SAGA covariates ---#
-#' create_covariates(dtm = tmp, layers = "all", output = dir, SAGApath = SAGApath)
+#' create_covariates(
+#'   dtm = fs::path(PEMprepr::read_fid()$dir_1020_covariates$path_rel,"25m","dem.tif",
```
```suggestion
#'   dtm = fs::path(PEMprepr::read_fid()$dir_1020_covariates$path_rel,"25m","dem.tif"),
```

gcperk commented:
Thanks @ateucher for the review. I updated the create_covariates to update default dem to sinksfilled and updates the csv alongside. I incorporated the minor fixes you found (thanks!). I will now merge back to main so that leaves the saga backend fix which you have on a separate branch.

ateucher commented:
That's perfect, thanks @gcperk - looks good!
